### PR TITLE
Fix DHCPCD --deprovision bug

### PR DESCRIPTION
**Pull request by Bug-Reaper**
## Description

Issue #1890 
https://github.com/Azure/WALinuxAgent/issues/1890

Fixes bug where `waagent --deprovision` would error/crash when trying to delete cached DHCPCD leases.

This error occurs because DHCPCD mounts files onto sub-directories of `/var/lib/dhcpcd`. These mounted DHCPCD files cannot be deleted whilst mounted (even with the all mighty power of `sudo rm`). This causes `waagent --deprovision` to crash when attempting to delete cached DHCPCD leases.

This PR adds the new method `umount(path)` to common/utils/fileutil.py. If the `path` argument does not pass os.path.ismount() then nothing happens. Otherwise the argument `path` is unmounted.

Right before we delete cached DHCPCD leases in the `del_dhcp_lease`, we call the method `fileutil.umount()` against paths of known DHCPCD mountpoints that cause this error. Because some mountpoints cannot be unmounted while DHCPCD is in use, we also first make sure that the DHCPCD service is not running. (pa/deprovision/default.py)

In doing so this PR address the crash/error and has been tested to work.

---

### PR information
- [ x ] The title of the PR is clear and informative.
- [ x ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ x ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ x ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

``` 
___      ___            __       ___      ___       __      
|"  |    |"  |          /""\     |"  \    /"  |     /""\     
||  |    ||  |         /    \     \   \  //   |    /    \    
|:  |    |:  |        /' /\  \    /\\  \/.    |   /' /\  \   
 \  |___  \  |___    //  __'  \  |: \.        |  //  __'  \  
( \_|:  \( \_|:  \  /   /  \\  \ |.  \    /:  | /   /  \\  \ 
 \_______)\_______)(___/    \___)|___|\__/|___|(___/    \___)                                                                                                                          
```

### CODE REVIEW

File: azurelinuxagent/common/utils/fileutil.py [hunk 1]
```diff
@@ -123,6 +124,15 @@ def chmod(path, mode):
         os.chmod(path, mode)
 
 
+def umount(*args):
+    for paths in args:
+        # find all possible mounts
+            for path in glob.glob(paths):
```
Comment (trstringer) [inline azurelinuxagent/common/utils/fileutil.py:130]: Unnecessary indentation here after the comment.

File: azurelinuxagent/common/utils/fileutil.py [hunk 2]
```diff
@@ -123,6 +124,15 @@ def chmod(path, mode):
...
+                if os.path.ismount(path):
+                    command=["umount",path]
+                    shellutil.run_command(command)
```
Comment (trstringer) [inline azurelinuxagent/common/utils/fileutil.py:133]: Is the usage of the `command` var necessary here? Looks like a single-use variable, consider combining this into a single line.

File: azurelinuxagent/pa/deprovision/default.py [hunk 3]
```diff
@@ -111,6 +111,14 @@ def del_resolv(self, warnings, actions):
 
     def del_dhcp_lease(self, warnings, actions):
         warnings.append("WARNING! Cached DHCP leases will be deleted.")
+
+        # For Arch based systems. Kill DHCPCD service. Needed to clear cached leases.
+        shellutil.run("systemctl stop dhcpcd", chk_err=False)
+
+        # For Arch based systems. Files mounted by DHCPCD can't be deleted whilst mounted, not even by root.
+        dirs_to_umount = ["/var/lib/dhcpcd/run/systemd/journal", "/var/lib/dhcpcd/run/udev", "/var/lib/dhcpcd/sys", "/var/lib/dhcpcd/dev", "/var/lib/dhcpcd/proc"]
+        actions.append(DeprovisionAction(fileutil.umount, dirs_to_umount))
```
Comment (trstringer) [inline azurelinuxagent/pa/deprovision/default.py:120]: If this is Arch-specific logic, then it should go into the `ArchDeprovisionHandler`.

**Review by trstringer — CHANGES_REQUESTED** [review]
Along with the individual suggestions, how are you testing this solution? What are your steps? Thanks for the contribution!

File: azurelinuxagent/common/utils/fileutil.py (hunk 2 above)
Comment (Bug-Reaper) [reply azurelinuxagent/common/utils/fileutil.py:133]: I agree, we can shorten this to one line.

File: azurelinuxagent/pa/deprovision/default.py (hunk 3 above)
Comment (Bug-Reaper) [reply azurelinuxagent/pa/deprovision/default.py:120]: So perhaps I should rephrase my comment here. While it's true this is for Arch systems, it's also needed for any *Nix distro that's running the latest version of `dhcpcd`. A quick check on wiki shows that a large number  of distros use `systemd/dhcpcd` by default or as an optional configuration:
Screenshot_2020-06-26_17-18-33


The systems that can be configured with `systemd/dhcpcd` include: Arch Linux, CentOS, CoreOS, Fedora, Mageia, Manjaro Linux, openSUSE, SUSE Linux Enterprise Server, Red Hat Enterprise Linux, Solus, Linux Mint, Ubuntu and Debian.


Among those, many use `systemd/dhcpcd` by default and with none/few alternatives.


I believe this is the correct place to run this, but I can update my comment to make it clear this is a DHCPCD specific issue and not an Arch one.

File: azurelinuxagent/common/utils/fileutil.py (hunk 1 above)
Comment (Bug-Reaper) [reply azurelinuxagent/common/utils/fileutil.py:130]: Yep, the indentation is unnecessary here. I'll amend that.

**Comment by Bug-Reaper** [comment]
Initial testing is covered in the issue.

Testing methodology for this PR has gone something like this:
(On a virtual system that's using the latest `systemd/dhcpcd`)
1. Install the latest waagent and run `waagent --deprovision` to confirm crash is still present on the latest version.
2. Clone my fork to my home directory.
3. Copy my forked version of the `azurelinuxagent` libs to replace the version installed by `waagent`.
4. Run `waagent --deprovision` again and confirm it completes successfully without crashing.
5. Shutdown my VM.
6. Upload the corresponding VHD file into Azure storage.
7. Create an Azure disk from that VHD file and deploy it as an Azure VM.
8. Confirm that VM is running normally (by SSHing in) and also that valid information is being reported about that VM in the Azure web portal.
